### PR TITLE
[BREAKING] Properly pad data from calldata.

### LINKED ISSUE #3493: Events emitted from `external` functions log different data than `public` functions
I posted full reproduction details on [StackExchange](https://ethereum.stackexchange.com/questions/39418/why-do-public-and-external-functions-emit-strings-differently-in-event-logs)

The short version is that calling these two functions produces different log output:

```solidity
pragma solidity 0.4.19;

contract EmitStringPublicAndExternal {
    event EmitString(string logme);

    function logPublic(string logme) public {
        EmitString(logme);
    }

    function logExternal(string logme) external {
        EmitString(logme);
    }
}
```

When I call both functions with "glitch" (in `geth --dev`), I get logs with two different data:

-  public: `0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000006676c697463680000000000000000000000000000000000000000000000000000`
-  external: `0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000006676c69746368`

_Note: "glitch" encoded with UTF-8, and then hex-encoded, is: `0x676c69746368`_

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #3493 

Note that as soon as this is merged, it is not possible to forward bytes data anymore via call. We need #3955 for this, but that is a breaking change.

### CODE REVIEW

**Comment by chriseth** [comment]
Actually #3955 can be implemented with `pragma 050`, so it might not be a big deal.

**Comment by axic** [comment]
Why is #3955 needed for that and why is that breaking? Is that the right issue to link to?

**Comment by chriseth** [comment]
`call()` performs padded encoding. This PR fixes a bug where the padding was not done for calldata byte arrays. Because of this bug, it was possible to forward call data via `x.call(msg.data)`. Once the bug is fixed, this is no longer possible. With https://github.com/ethereum/solidity/issues/3955#issuecomment-383539560 the idea is to change the `call` function so that it only accepts a single `bytes` argument which is encoded as is without padding.

**Comment by chriseth** [comment]
Result of discussion: Add a warning for hash functions, and do the rest (including call*) with 0.5.0 proper.

**Comment by chriseth** [comment]
Ready for review.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -4010,15 +4010,17 @@ BOOST_AUTO_TEST_CASE(call_forward_bytes_length)
 	compileAndRun(sourceCode, 0, "sender");
 
 	// No additional data, just function selector
-	ABI_CHECK(callContractFunction("viaCalldata()"), encodeArgs(4));
+	ABI_CHECK(callContractFunction("viaCalldata()"), encodeArgs(0x20));
+	// Should be this with 0.5.0: encodeArgs(4));
 	ABI_CHECK(callContractFunction("viaMemory()"), encodeArgs(0x20));
 	// Should be this with 0.5.0: encodeArgs(4));
 	ABI_CHECK(callContractFunction("viaStorage()"), encodeArgs(0x20));
```
Comment (axic) [inline test/libsolidity/SolidityEndToEndTest.cpp:3971]: Aren't these need to return 4?

Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp:3971]: No, this will be done in the other pull request. This is the pull request that adds padding for encoding from calldata to be identical to the encoding from all other types (this applies to `call` and to `abi.encodePacked`). There is another pull request that changes the behaviour of `call` to not use padding for anything.

**Comment by chriseth** [comment]
Rebased. @axic can this be merged?

File: libsolidity/codegen/ArrayUtils.cpp [hunk 2]
```diff
@@ -303,12 +303,16 @@ void ArrayUtils::copyArrayToMemory(ArrayType const& _sourceType, bool _padToWord
...
 		if (baseSize > 1)
 			m_context << u256(baseSize) << Instruction::MUL;
-		// stack: target source_offset source_len
-		m_context << Instruction::DUP1 << Instruction::DUP3 << Instruction::DUP5;
-		// stack: target source_offset source_len source_len source_offset target
-		m_context << Instruction::CALLDATACOPY;
-		m_context << Instruction::DUP3 << Instruction::ADD;
-		m_context << Instruction::SWAP2 << Instruction::POP << Instruction::POP;
+
+		string routine = "calldatacopy(target, source, len)\n";
```
Comment (axic) [inline libsolidity/codegen/ArrayUtils.cpp:307]: While I am fine with this, why not use Whiskers? Probably a bit overkill :)

File: libsolidity/codegen/ArrayUtils.cpp [hunk 3]
```diff
@@ -303,12 +303,16 @@ void ArrayUtils::copyArrayToMemory(ArrayType const& _sourceType, bool _padToWord
...
+		if (_padToWordBoundaries)
+			routine += R"(
+				mstore(add(target, len), 0)
```
Comment (axic) [inline libsolidity/codegen/ArrayUtils.cpp]: Why is this needed?

File: libsolidity/codegen/ArrayUtils.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ArrayUtils.cpp:307]: Yes, because of the overkill-factor.

File: libsolidity/codegen/ArrayUtils.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ArrayUtils.cpp]: To make sure that it actually ends in a sequence of zeros in case the memory was dirty.

Comment (axic) [reply libsolidity/codegen/ArrayUtils.cpp]: For a moment I thought this was mstore8 and that didn't make sense in terms of cleaning the entire slot. Still I'd think a short comment here wouldn't hurt explaining the intentions.

**Comment by axic** [comment]
Rebased.

**Review by axic — APPROVED** [review]
